gpjax/likelihoods.py

```diff
@@ -199,6 +199,44 @@ def predict(self, dist: tfd.Distribution) -> tfd.Distribution:
         return self.link_function(mean / jnp.sqrt(1.0 + variance))
 
 
+@dataclass
+class Beta(AbstractLikelihood):
+    def link_function(self, f: Float[Array, "..."]) -> tfd.Distribution:
+        r"""The probit link function of the Beta likelihood.
+
+        Args:
+            f (Float[Array, "..."]): Function values.
+
+        Returns
+        -------
+            tfd.Distribution: The likelihood function.
+        """
+
+        return tfd.Beta(inv_probit(f[0]), inv_probit(f[1]))
```

Comment:
This is a step in the right direction, but note our variable `f` is vector of length $N$. In particular, this is vector of function values $\mathbf{f} = (f_1, \dotsc, f_N) \in \mathbb{R}^{N}$.

As such our Beta likelihood, $p(\mathbf{y}|\mathbf{f})$  models observations $\mathbf{y} = (y_1, \dotsc, y_N)$, that are drawn $y_i \sim \text{Beta}(\alpha_i, \beta_i)$ with $\mathbf{\alpha} = \nu \Phi (\mathbf{f})$, $\mathbf{\beta} = \nu (1 - \Phi (\mathbf{f}))$, $\mathbf{\alpha} = (\alpha_1, \dotsc, \alpha_N)$, $\mathbf{\beta} = (\beta_1, \dotsc, \beta_N)$. 

You need to pass these $\alpha$ and $\beta$ vectors to `tfd.Beta`.

For the dispersion parameter $\nu$, you'll need to mimic the `obs_noise` variable for our `Gaussian` likelihood.

Reply:
Hey @brandall25,

Thanks, good questions. I'll answer these in two separate comments.

> Something like this? I'm not sure f is needed in the link function.

Ah really we mean an inverse link function here, but we have rather relaxed this term.

We are looking for  for the map

$$f_i \mapsto p(y_i | g^{-1}(f_i))$$ 

where $g(\mu_i) = f_i$ and $\mu_i$ is the mean of the beta distribution, $g$ is the *link function* in the context of Generalised linear models, $y_i$ are the observations. So in particular, we are looking to reparameterise your code:

```python
def link_function(self, alpha: Float[Array, "..."], beta: Float[Array, "..."]) -> tfd.Distribution:
        r"""The probit link function of the Beta likelihood.
       ...

```

In terms of $\mathbf{f}$, as 

 ```python
def link_function(self, f: Float[Array, "N"]) -> tfd.Distribution:
        r"""The probit link function of the Beta likelihood.
       ...
```
And we need to use something to relate $\mathbf{f}$ to $\mathbf{\alpha}$ and $\mathbf{\beta}$ of the beta distribution. We suggested to do this with the $\Phi$ and dispersion parameter $\nu$.

Reply:
> Also, I'm not understanding what you mean by mimicking obs_noise for the dispersion parameter. You're saying v should be defined as a union like obs_noise`? Why is that?

Regarding the `obs_noise` term on the `Gaussian` likelihood - notice this is a special type of dataclass field, called a `param_field` as shown below: 

```python
@dataclass
class Gaussian(AbstractLikelihood):
    r"""Gaussian likelihood object."""

    obs_noise: Union[ScalarFloat, Float[Array, "#N"]] = param_field(
        jnp.array(1.0), bijector=tfb.Softplus()
    )
    ...
```

As such the `param_field` defines two things:

- (1) The default value the parameter should take, in this instance `jnp.array(1.0)` 
- (2) The domain  of the parameter through `bijector=tfb.Softplus()` - meaning this is a positive parameter. 

For the beta likelihood, we would need to have something that takes this same structure for our `dispersion` parameter $\nu$, as that too is a positive parameter, and you would need to do something like this.
```python
@dataclass
class Beta(AbstractLikelihood):
    r"""Beta likelihood object."""

    dispersion: Union[ScalarFloat, Float[Array, "#N"]] = param_field(
        jnp.array(1.0), bijector=tfb.Softplus()
    )
    ...
```

Reply:
So is `f` is the pdf of `dist`? 

It would be called from `predict` something like this `return self.link_function(dist.prob())`

And then `alpha` and `beta` are derived from `f` in `link_function`?

Reply:
Hi @brandall25,

This is not quite the case no. I'll give a bit more detail on this below:

## --- Details ---

### `link_function`
- `f` parameterises the pdf of $y$ and this is what we want to return. This object is the conditional distribution $y | f$. 
- Specifically , we want to return the `tfd.Beta(alpha, beta)`, where `alpha = self.dispersion * inv_probit(f)` and `beta = self.dispersion * (1 - inv_probit(f))`. 

### `predict `
- The `dist` in  `predict` method will be an arbitrary Multivariate Gaussian $\mathbf{f}= (f_1, \dotsc , f_n) \sim N(\mu, \Sigma)$ where $\mu$ is a mean for $\mathbf{f}$ and $\Sigma$ is the covariance.  (This will actually be the predictive distribution for the function of interest `f`).

- To do the `predict`, you will need to draw samples from `dist`, $\mathbf{f}^{(j)} \sim  N(\mu, \Sigma)$, $j=1,..., M$, and return a Monte Carlo estimate -> this is a Mixture distribution $\mathbf{y}^* \sim \frac{1}{M} \sum_{j} p(\cdot| \mathbf{f}^{j}) $, which can be implemented with `tfd.Mixture` (see, https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Mixture) . 
- As such,  `tfd.Mixture` has two arguments;  `components` that will be a list of `link_function`s evaluated at each $\mathbf{f}^{j}$ (so i.e., a list of the corresponding `tfd.Distribution` that each `link_function`($\mathbf{f}^{j}$) gives). And the other argument is a `cat` argument which should be uniform $\frac{1}{M}$ weights. 
- We actually need this `predict` method in general. So it would be nice to have this on the `AbstractLikelihood` class, to serve as the default option.
 
## --- Takeaway ---

- I strongly recommend focusing on the `link_function` for now, getting this coded in would be an excellent start. 
- Then, if you feel comfortable having a crack at the `predict` method, then please do so! Also I'm happy to help with it, just give me a shout.


Cheers,
Dan